File: Backend/src/main/java/com/espumablanca/backend/application/usecase/AuthService.java
package com.espumablanca.backend.application.usecase;

import com.espumablanca.backend.domain.model.dto.request.auth.AuthLoginRequest;
import com.espumablanca.backend.domain.model.dto.response.auth.AuthDto;
import com.espumablanca.backend.domain.model.dto.response.auth.AuthLoggedDto;
import com.espumablanca.backend.domain.model.dto.response.auth.RefreshTokenDto;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

public interface AuthService {

    AuthDto login(AuthLoginRequest request);

    RefreshTokenDto refreshToken(HttpServletResponse response, HttpServletRequest request);

    AuthLoggedDto currentUser();

    void validateCookies(HttpServletResponse response, HttpServletRequest request);

    void logout(HttpServletResponse response, HttpServletRequest request);
}

File: EspublaBlancaFrontend/src/app/core/services/api/auth.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { switchMap, tap } from 'rxjs';
import { environment } from '../../../../environments/environment.development';
import { AuthLoginRequest } from '../../models/auth/request/auth-login-request.model';
import { CurrentUser } from '../../models/auth/response/current-user.model';
import { AuthorityStorageService } from '../application/authority-storage.service';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  private readonly endpoint = environment.apiUrl + '/public/auth';
  private readonly http = inject(HttpClient);
  private readonly authorityStorage = inject(AuthorityStorageService);

  login(request: AuthLoginRequest) {
    return this.http.post<void>(this.endpoint, request).pipe(
      switchMap(() => {
        return this.currentUser();
      })
    );
  }

  logout() {
    return this.http.post<void>(this.endpoint + '/logout', {}).pipe(
      tap(() => {
        this.authorityStorage.updateCurrentUser({} as CurrentUser);
      })
    );
  }

  currentUser() {
    return this.http.get<CurrentUser>(this.endpoint + '/current-user').pipe(
      tap((resp) => {
        this.authorityStorage.updateCurrentUser(resp);
      })
    );
  }

  refreshToken() {
    return this.http.post<void>(this.endpoint + '/refresh-token', {}).pipe(
      switchMap(() => {
        return this.currentUser();
      })
    );
  }
}
